### ISSUE TITLE
feat: optimize GM prompts against evaluation metrics

### ISSUE THREAD

**Issue by omegafrog** [omegafrog/dnd-master#210]
## 목표

GM 시스템의 prompt를 사람이 감으로 반복 수정하는 방식에서 벗어나, **GM Eval 결과를 목적 함수로 사용해 planner/writer/judge/verifier prompt를 체계적으로 비교·최적화하는 평가 기반 prompt optimization 체계**를 도입한다.

이번 이슈는 우선순위 8번인 **Prompt Optimization**만 다룬다.

---

## 배경

프롬프트를 수동으로 수정하면 다음 문제가 발생한다.

- 특정 예제에는 좋아지지만 다른 장면에서 악화됨
- 어떤 수정이 실제 품질 향상에 기여했는지 알기 어려움
- 모델 변경 시 prompt를 다시 처음부터 조정해야 함
- planner / writer / verifier 각각의 역할에 다른 최적화 기준이 필요한데 하나의 prompt 감각에 의존함
- 회귀(regression)를 놓치기 쉬움

따라서 prompt를 code/config artifact로 취급하고, 평가 데이터셋에 대해 반복적으로 측정 가능한 optimization 대상으로 만든다.

---

### 1. Eval이 먼저다

최적화는 반드시 고정된 GM Eval suite와 metric 위에서 수행한다.

```text
Prompt Candidate
      ↓
Eval Dataset
      ↓
Metrics
      ↓
Comparison
      ↓
Selected Prompt
```

평가 기준 없이 자동 prompt search부터 도입하지 않는다.

### 2. 역할별 prompt를 별도로 최적화한다

예:

```text
Turn Planner prompt
→ planning validity / agency / story progress

Writer prompt
→ prose / tone / factual faithfulness

Judge prompt
→ preference consistency

Verifier prompt
→ violation detection precision/recall
```

### 3. production prompt는 versioned artifact다

모델 호출 코드 안에 긴 문자열로 흩어놓지 않고 다음 메타데이터를 추적할 수 있어야 한다.

```text
prompt id
prompt version
role
model
optimization run
metric result
created/selected reason
```

---

## 최적화 대상

Prompt 자체뿐 아니라 필요하면 다음 항목도 하나의 program configuration으로 취급한다.

```text
system instruction
role instruction
few-shot examples
output schema description
rubric wording
context ordering
retrieved exemplar placement
```

단, model weights는 이번 이슈 범위가 아니다.

---

## 최적화 흐름

```text
Baseline Prompt
      ↓
Candidate Generation / Mutation
      ↓
Offline Eval
      ↓
Hard Constraint Filter
      ↓
Quality Ranking
      ↓
Best Candidate
      ↓
Regression Check
      ↓
Prompt Registry
```

DSPy의 optimizer/teleprompter 계열이나 GEPA/MIPRO 류 접근을 검토할 수 있지만, 특정 프레임워크에 도메인을 강하게 결합하지 않도록 adapter boundary를 둔다.

---

## Metric 구성

단일 평균 점수만 최대화하지 않는다.

예:

```text
Hard Metrics
- rule violation rate
- secret leak rate
- player agency violation rate
- schema failure rate

Soft Metrics
- coherence
- NPC consistency
- pacing
- interestingness
- prose quality
```

Hard constraint 악화를 soft score 개선으로 상쇄하지 못하게 한다.

예:

```text
secret leak > baseline
→ reject candidate

agency violation > threshold
→ reject candidate

otherwise
→ compare quality score
```

---

## Train / Dev / Regression 분리

Prompt optimization이 평가 세트에 과적합되지 않도록 데이터셋을 분리한다.

```text
optimization/train set
→ candidate search

dev set
→ candidate selection

regression/holdout set
→ 최종 확인
```

같은 몇 개의 장면을 계속 보고 prompt를 수정하는 방식은 피한다.

---

### Planner

목표:

- valid TurnPlan 생성률
- player intent 보존
- 불필요한 rule request 감소
- Story Plan과의 일관성
- meaningful consequence

### Writer

목표:

- TurnPlan fidelity
- secret safety
- 자연스러운 prose
- 상황에 맞는 tone/pacing
- player agency 유지

### Verifier

목표:

- 실제 오류 탐지율
- 정상 응답 false positive 감소
- rewrite 지시의 정확성

### Judge

목표:

- human/preferred ranking과 일치
- candidate 순서 변경에도 안정적
- 동일 후보 재평가 시 일관성

---

## Prompt Registry

최소 다음 구조를 제공한다.

예:

```json
{
  "promptId": "gm-writer",
  "version": "v12",
  "model": "...",
  "parentVersion": "v11",
  "optimizationRunId": "opt-2026-08-01",
  "metrics": {
    "agencyViolationRate": 0.01,
    "secretLeakRate": 0.0,
    "qualityScore": 0.87
  }
}
```

production에서 어떤 prompt/model 조합으로 응답이 생성됐는지 재현할 수 있어야 한다.

---

## 자동 최적화와 수동 검토

자동 metric 1위 prompt를 즉시 production에 반영하지 않는다.

권장 흐름:

```text
optimizer selects candidates
      ↓
offline metrics
      ↓
representative sample inspection
      ↓
regression suite
      ↓
approve
```

특히 creative writing metric은 완전한 객관식 점수가 아니므로 대표 출력 검토를 남긴다.

---

### Case 1. Baseline 비교

현재 production prompt가 baseline으로 저장되고 동일 eval suite에서 candidate와 비교된다.

### Case 2. 회귀 차단

새 prompt가 prose score는 높지만 secret leak이 증가한다.

- 최종 candidate로 선택되지 않는다.

### Case 3. 역할별 분리

Writer prompt 개선이 Planner prompt를 자동으로 덮어쓰지 않는다.

- 각 role의 prompt/version이 독립 관리된다.

### Case 4. 재현성

특정 optimization run의 dataset/version/model/prompt 후보와 결과를 다시 확인할 수 있다.

### Case 5. Holdout

optimization에 사용되지 않은 regression/holdout set에서 최종 성능을 측정한다.

### Case 6. Rollback

새 prompt에서 production 문제 발생 시 이전 version으로 되돌릴 수 있다.

---

## 관측성

최소 기록:

```text
optimization run id
prompt versions
model versions
dataset version
candidate metrics
selected candidate
baseline delta
hard constraint regressions
```

---

## 이번 이슈에서 제외

- GM Eval metric 자체의 최초 설계
- TurnPlan 설계
- Narrative State
- Best-of-N runtime planning
- Style Exemplar Retriever 자체 구현
- model fine-tuning / LoRA / preference tuning
- online reinforcement learning

이번 이슈의 목적은 **프롬프트를 감각적으로 다듬는 작업에서 벗어나, 고정된 평가 기준을 통해 재현 가능하고 회귀를 통제할 수 있는 최적화 프로세스**를 만드는 것이다.

**Comment by omegafrog** [comment]
Approved implementation slices:
- #231 GMQ-001 Prompt Registry / baseline
- #232 GMQ-002 gated candidate evaluation
- #233 GMQ-003 approval, activation, rollback

Dependency: #220 → #231 → #232 → #233.
